src/software/amazon/ion/impl/lite/IonIntLite.java

```diff
@@ -34,11 +35,18 @@
     static private final BigInteger LONG_MAX_VALUE =
         BigInteger.valueOf(Long.MAX_VALUE);
 
+    static private final BigInteger INT_MIN_VALUE =
+        BigInteger.valueOf(Integer.MIN_VALUE);
+
+    static private final BigInteger INT_MAX_VALUE =
+        BigInteger.valueOf(Integer.MAX_VALUE);
+
     private static final int HASH_SIGNATURE =
         IonType.INT.toString().hashCode();
 
     private long _long_value;
     private BigInteger _big_int_value;
+    private IntegerType _actual_type;
```

Comment:
One of the main motivations behind this feature is that BigIntegers are really expensive to construct compared to primitives.

It's convenient to memoize `_actual_type` when the value is set, because all necessary code paths are hit. After refactoring to make this a callback API, there's no longer an `IntegerType` enum, so I replaced it in here with a byte to hold the type info.